File: clients/client-js/src/transport/sse-transport.ts
import { MessageDecoder, JsonDecoder } from "../decoder/index.js"
import { ChannelMessage } from "../channel-message.js";
import { Transport } from "./transport.js";
import { AsyncConfig } from "../async-config.js";
import { EventSourceController, EventSourcePlus } from "event-source-plus";
import { TransportError } from "./transport-error.js";
import { Utils } from "../utils.js";

export class SseTransport implements Transport {

    private actualToken;
    private readonly serializer: MessageDecoder;
    private errorCount = 0;
    private controller: EventSourceController;

    private static readonly MAX_RETRY_INTERVAL = 6000;
    private static readonly SUCESS_STATUS = 299;

    public static create(config: AsyncConfig,
        handleMessage = (_message: ChannelMessage) => { },
        errorCallback = (_error: TransportError) => { },
        transport: EventSourcePlus = null): Transport {
            return new SseTransport(config, handleMessage, errorCallback, transport);
        }

    private constructor(private readonly config: AsyncConfig,
        private readonly handleMessage: (_message: ChannelMessage) => void,
        private readonly errorCallback: (_error: TransportError) => void,
        private readonly transport:EventSourcePlus = null) {
        this.serializer = new JsonDecoder();
        this.actualToken = config.channel_secret;
        this.transport = transport || new EventSourcePlus(this.sseUrl(), {
            maxRetryInterval: SseTransport.MAX_RETRY_INTERVAL,
            maxRetryCount: this.config.maxReconnectAttempts,
        });
        console.debug('async-client. sse transport created with transport: ', transport);
    }

    name(): string {
        return 'sse';
    }

    public connect(connectedCallback?: () => void): void {
        if (this.connected()) {
            console.debug('async-client. sse already created and open');
            return;
        } else {
            console.debug('async-client. sse not connected, creating new EventSource');
        }

        this.controller = this.transport.listen({
            onMessage: (event) => {             
                try {
                    const message = this.serializer.decode_sse(event.data)
                    if (message.event == ":n_token") {
                        this.actualToken = message.payload;
                        console.debug('async-client. sse received new token = ', this.actualToken);
                    }
                    this.errorCount = 0;
                    this.handleMessage(message);
                } catch (error) {
                    console.error('Error processing message:', error);
                }
            },
            onRequest: ({ options }) => {
                options.headers.append("Authorization", "Bearer " + this.getToken());
            },
            async onResponse({ response }) {
                console.debug(`Sse client received status code: ${response.status}`);
                if (connectedCallback && response.status <= SseTransport.SUCESS_STATUS) {
                    connectedCallback();
                }
            },
            async onResponseError({ request, response, error }) {
                this.errorCount++;
                console.debug(`[Sse response error]`, request, response.status, error);
                const body = await response.text();
                console.error('Sse response error:', body);
                let parsed;
                try {
                    parsed = JSON.parse(body);
                } catch (e) {
                    console.error('Error parsing response:', e);
                    parsed = { error: body };
                }
                const reason = Utils.extractReason(parsed.error);
                const stopRetries = response.status == 400 ||
                    response.status == 404 ||
                    (response.status == 401 && reason < 3050) ||
                    (response.status == 428 && reason < 3050);

                if (stopRetries || this.errorCount >= this.config.maxReconnectAttempts) {
                    console.log('async-client. sse stopping retries');
                    this.disconnect();
                    this.errorCallback({ origin: 'sse', code: 1, message: response.statusText + ' ' + body });
                } else if (response.status === 401) {
                    console.log('async-client. disconnecting because 401 and will retry with new token');
                    this.disconnect();
                    this.connect(connectedCallback);
                }
            },

            onRequestError(context) {
                console.error('Sse request error:', context.error.message);
            },
        });

        console.debug('async-client. sse connect() called')
    }

    public disconnect(): void {
        console.info('async-client. sse disconnect() called');
        if (!this.controller.signal.aborted) {
            this.controller.abort();
            console.debug('async-client. sse aborted');
        } else {
            console.debug('async-client. sse already aborted');
        }
    }

    public connected(): boolean {
        return this.transport && this.controller && !this.controller.signal.aborted;
    }

    public send(message: string): void {
        console.warn('async-client. sse transport does not support sending messages. Use the WebSocket transport for sending messages.');
    }

    // only for testing or internal
    public getDecoder(): MessageDecoder {
        return this.serializer;
    }

    private getToken(): string {
        return this.actualToken;
    }

    private sseUrl(): string {
        if (this.config.sse_url) {
            return `${this.config.sse_url}/ext/sse?channel=${this.config.channel_ref}`;
        }
        console.debug('async-client. sse config.sse_url not set, trying to compute from socket_url');
        // try to compute the SSE URL from the socket URL
        let url = this.config.socket_url;
        if (url.startsWith('ws')) {
            url = url.replace('ws', 'http');
        } else if (url.startsWith('wss')) {
            url = url.replace('wss', 'https');
        }
        return `${url}/ext/sse?channel=${this.config.channel_ref}`;
    }
}

File: clients/client-dart/lib/src/transport/types/sse_transport.dart
// ignore_for_file: avoid-ignoring-return-values
import 'dart:async';

import 'package:logging/logging.dart';

import '../../../channel_sender_client.dart';
import '../../decoder/decoder.dart';
import '../../exceptions/exceptions.dart';
import '../../sse/eventflux.dart';
import '../../utils/capped_list.dart';
import '../../utils/retry_timer.dart';
import '../transport.dart';

class SSETransport implements Transport {
  MessageDecoder msgDecoder = JsonDecoder();
  static const String EVENT_KIND_USER = 'user_event';
  static const String EVENT_KIND_SYSTEM = 'system_event';
  static const String RESPONSE_NEW_TOKEN = ':n_token';

  static const int RETRY_DEFAULT_MAX_RETRIES = 5;
  static const int RETRY_DEFAULT_MAX_TIME = 2000;
  static const int RETRY_DEFAULT_MIN_TIME = 350;

  static const int SSE_OK_CLOSE_CODE = 200;

  late String currentToken;

  final _log = Logger('SSETransport');

  final void Function(int, String) _signalSSEClose;
  final void Function(Object) _signalSSEError;
  final AsyncConfig _config;
  late RetryTimer _connectRetryTimer;

  late StreamController<ChannelMessage> _broadCastStream;
  final CappedList<String> _messageDedup = CappedList<String>(50);

  SSETransport(
    this._signalSSEClose,
    this._signalSSEError,
    this._config,
  ) {
    currentToken = _config.channelSecret;
    _broadCastStream = StreamController<
        ChannelMessage>.broadcast(); // subscribers stream of data
    _connectRetryTimer = RetryTimer(
      () async {
        return await connect();
      },
      () async {
        _signalSSEError(
          MaxRetriesException(
            '[async-client][SSETransport] Max retries reached',
          ),
        );
      },
      initialWait: RETRY_DEFAULT_MIN_TIME,
      maxWait: RETRY_DEFAULT_MAX_TIME,
      maxRetries: _config.maxRetries,
    );
  }

  @override
  TransportType name() {
    return TransportType.sse;
  }

  @override
  Future<bool> connect() async {
    _log.finer('[async-client][SSETransport] connect() started.');

    EventFlux.instance.connect(
      EventFluxConnectionType.get,
      sseUrl(),
      header: {
        'Authorization': 'Bearer $currentToken',
        'Accept': 'text/event-stream',
      },
      multipartRequest: false,
      onSuccessCallback: (EventFluxResponse? response) {
        response?.stream?.listen(
          (data) {
            _onData(data.data);
          },
        );
      },
      onError: (error) {
        _onResponseError(error, StackTrace.current);
      },
      onConnectionClose: () {
        _log.warning('[async-client][SSETransport] onConnectionClose called');
        _signalSSEClose(0, '');
      },
      autoReconnect: true, // Keep the party going, automatically!
      reconnectConfig: ReconnectConfig(
        mode: ReconnectMode.linear, // or exponential,
        interval: const Duration(seconds: 2),
        maxAttempts: _config.maxRetries ??
            RETRY_DEFAULT_MAX_RETRIES, // or -1 for infinite,
        onReconnect: () {
          _log.info('[async-client][SSETransport] onReconnect Called');
        },
        reconnectHeader: () => Future.value({
          'Authorization': 'Bearer $currentToken',
          'Accept': 'text/event-stream',
        }),
      ),
    );

    _log.finest('[async-client][SSETransport] connect() called');

    return true;
  }

  @override
  void send(String message) {
    _log.warning('[async-client][SSETransport] method Send not supported');
  }

  void _handleNewToken(ChannelMessage message) {
    _log.finest('[async-client][SSETransport] new token received');
    currentToken = message.payload;
    _config.channelSecret = currentToken;
  }

  void _onResponseError(error, stackTrace) {
    EventFluxException parsedException;

    parsedException = error is EventFluxException
        ? error
        : EventFluxException(
            message: error?.toString() ?? 'Unknown SSE error',
            statusCode: null,
            reasonPhrase: null,
          );

    // Provide default values for null status codes and reason phrases
    int? statusCode = parsedException.statusCode;
    String? reasonPhrase = parsedException.reasonPhrase ?? 'Unknown error';

    // close code 200 is ok to ignore
    if (statusCode != SSE_OK_CLOSE_CODE) {
      _log.severe(
        '[async-client][SSETransport] Error in SSE connection: ${parsedException.statusCode}, ${parsedException.reasonPhrase}, stackTrace: $stackTrace',
      );
      // ignore: prefer-async-await
      EventFlux.instance.disconnect().then(
        (_) {
          if (!_connectRetryTimer.isActive()) {
            _connectRetryTimer.schedule();
          } else {
            _log.warning(
              '[async-client][SSETransport] Retry timer is already active. Waiting for it to finish.',
            );
          }
        },
        onError: (e) {
          _log.severe(
            '[async-client][SSETransport] Error calling EventFlux.disconnect(): $e, stackTrace: $stackTrace',
          );
        },
      );
    }
  }

  int extractCode(String stringCode) {
    return int.tryParse(stringCode) ?? 0;
  }

  void _onData(String data) {
    _log.finest('[async-client][SSETransport] Received raw from Server: $data');
    var message = msgDecoder.decode(data);

    if (message.event == RESPONSE_NEW_TOKEN) {
      _handleNewToken(message);
    }

    if (_messageDedup.contains(message.messageId ?? '')) {
      _log.warning(
        '[async-client][SSETransport] message deduped: ${message.messageId}',
      );

      return;
    } else {
      _messageDedup.add(message.messageId ?? '');
      _broadCastStream.add(message);
    }
  }

  String sseUrl() {
    String url = '';
    if (_config.sseUrl != null) {
      url = _config.sseUrl ?? '';
      url = '$url?channel=${_config.channelRef}';
      _log.info('[async-client][SSETransport] url is $url');
    } else {
      if (_config.socketUrl.startsWith('ws')) {
        url = _config.socketUrl
            .replaceFirstMapped('ws:', (match) => 'http:')
            .replaceFirstMapped('wss:', (match) => 'https:')
            .replaceFirstMapped('/ws', (match) => '/sse')
            .replaceFirstMapped('/socket', (match) => '/sse');
      }
      url = '$url?channel=${_config.channelRef}';
      _log.info('[async-client][SSETransport] Calculated url will be $url');
    }

    return url;
  }

  @override
  bool isOpen() => false;

  @override
  Stream<ChannelMessage> get stream => _broadCastStream.stream;

  @override
  Future<void> disconnect() async {
    _log.info('[async-client][SSETransport] disconnect() called');
    await EventFlux.instance.disconnect();
    _connectRetryTimer.reset();
    _log.info('[async-client][SSETransport] disconnect() finished');

    return;
  }
}
